### ISSUE TITLE
StackOverflowError with Test.detect_ambiguities

### ISSUE THREAD

**Issue by rschwarz** [JuliaLang/julia#35909]
I'm getting a `StackOverflowError` when I call `Test.detect_ambiguities` with multiple packages, e.g. `SparseArrays` and `Base`. But this only happens after `using` certain other packages.

This was originally discovered in https://github.com/tkf/Aqua.jl/issues/16 with one of my own (unregistered) packages, but I was able to reproduce it in a simpler case (with lighter dependencies).

See [this gist](https://gist.github.com/rschwarz/a0368d51ea7d9b458e64e3d7f00a0c5c) for MWE with output from Julia 1.4.0.

**Comment by rschwarz** [comment]
I justed tested again on Julia 1.4.1 with the same behavior. In short, this is the problem:
```julia
julia> Test.detect_ambiguities(SparseArrays, Base; recursive=false)
0-element Array{Tuple{Method,Method},1}

julia> using MathOptInterface

julia> Test.detect_ambiguities(SparseArrays, Base; recursive=false)
ERROR: StackOverflowError:
Stacktrace:
 [1] _methods_by_ftype at ./reflection.jl:841 [inlined]
 [2] _methods_by_ftype(::Any, ::Int64, ::UInt64) at ./reflection.jl:838
 [3] isambiguous(::Method, ::Method; ambiguous_bottom::Bool) at ./reflection.jl:1301
 [4] detect_ambiguities(::Module, ::Vararg{Module,N} where N; imported::Bool, recursive::Bool, ambiguous_bottom::Bool) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1440
 [5] top-level scope at REPL[11]:1
```

**Comment by JeffBezanson** [comment]
Looks like the same sort of thing as #35269.

**Comment by sbromberger** [comment]
I don't know whether this is the same issue, but doing a `isempty(detect_unbound_args(LightGraphs, recursive=true))`

results in
```
Skipping LightGraphs.Experimental.ShortestPaths                    
Skipping LightGraphs.Experimental.SubgraphIsomorphismProblem       
Skipping LightGraphs.Experimental.InducedSubgraphIsomorphismProblem
Skipping LightGraphs.Experimental.ShortestPaths                                                                                                       
Skipping LightGraphs.Experimental.SubgraphIsomorphismProblem
```
until a `StackOverFlowError`:
```
Skipping LightGraphs.ExperimentalERROR: StackOverflowError:
Stacktrace:
 [1] poptaskref(::Base.InvasiveLinkedListSynchronized{Task}) at ./task.jl:702
 [2] wait at ./task.jl:709 [inlined]
 [3] uv_write(::Base.TTY, ::Ptr{UInt8}, ::UInt64) at ./stream.jl:935
...
```

**Comment by thofma** [comment]
The error from OP is fixed in master (more precisely, by 258d04e8bfdb8189fd61199bccd7641650fec07b)

**Comment by rschwarz** [comment]
Given the above comment, I'm happy to close this.
